The job **`x86_64-gnu-nopt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12880527386/job/35909709431) [(plain)](https://github.com/rust-lang-ci/rust/commit/663dd5e49ad0a60b62a4c74236c1d78a48de529a/checks/35909709431/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/slice-init.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll" "/checkout/tests/codegen/slice-init.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/slice-init.rs:88:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/slice-init.rs:88:12: error: CHECK: expected string not found in input
 // CHECK: call void @llvm.memset.p0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll:97:30: note: scanning from here
define void @option_none_init(ptr sret([200 x i8]) align 1 %_0) unnamed_addr #0 {
                             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll:126:3: note: possible intended match here
define void @option_maybe_uninit_init(ptr sret([200 x i8]) align 2 %_0) unnamed_addr #0 {

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/slice-init/slice-init.ll
Check file: /checkout/tests/codegen/slice-init.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'slice_init.587352d4ad61cb6b-cgu.0' 
            2: source_filename = "slice_init.587352d4ad61cb6b-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            5:  
            5:  
            6: %"core::option::Option<u8>" = type { i8, [1 x i8] } 
            7: %"core::mem::maybe_uninit::MaybeUninit<Packed>" = type { [12 x i8] } 
            8:  
            9: @0 = private unnamed_addr constant <{ [1 x i8], [1 x i8] }> <{ [1 x i8] zeroinitializer, [1 x i8] undef }>, align 2 
           10: @1 = private unnamed_addr constant <{ ptr, [4 x i8] }> <{ ptr @"_ZN10slice_init30option_maybe_uninit_provenance29_$u7b$$u7b$constant$u7d$$u7d$8HAS_ADDR17h48e24edc0f910752E", [4 x i8] undef }>, align 1 
           11: @"_ZN10slice_init30option_maybe_uninit_provenance29_$u7b$$u7b$constant$u7d$$u7d$8HAS_ADDR17h48e24edc0f910752E" = internal constant <{}> zeroinitializer, align 1 
           12:  
           13: ; Function Attrs: nonlazybind uwtable 
           14: define void @zero_sized_elem() unnamed_addr #0 { 
           16:  %_1 = alloca [0 x i8], align 1 
           17: ; call slice_init::opaque 
           17: ; call slice_init::opaque 
           18:  call void @_ZN10slice_init6opaque17h9c01bc0e6568e7f7E(ptr align 1 %_1) 
           19:  ret void 
           20: } 
           21:  
           22: ; Function Attrs: nonlazybind uwtable 
           23: define void @zero_len_array() unnamed_addr #0 { 
           25:  %_1 = alloca [0 x i8], align 4 
           26: ; call slice_init::opaque 
           26: ; call slice_init::opaque 
           27:  call void @_ZN10slice_init6opaque17h90ba94c9c1d53e83E(ptr align 4 %_1) 
           28:  ret void 
           29: } 
           30:  
           31: ; Function Attrs: nonlazybind uwtable 
           32: define void @byte_array() unnamed_addr #0 { 
           33: start: 
           34:  %x = alloca [4 x i8], align 1 
           35:  call void @llvm.memset.p0.i64(ptr align 1 %x, i8 7, i64 4, i1 false) 
           36: ; call slice_init::opaque 
           37:  call void @_ZN10slice_init6opaque17h5a1c3b8d7de917daE(ptr align 1 %x) 
           38:  ret void 
           39: } 
           40:  
           41: ; Function Attrs: nonlazybind uwtable 
           42: define void @byte_enum_array() unnamed_addr #0 { 
           44:  %_2 = alloca [1 x i8], align 1 
           44:  %_2 = alloca [1 x i8], align 1 
           45:  %x = alloca [4 x i8], align 1 
           46:  store i8 1, ptr %_2, align 1 
           47:  %0 = load i8, ptr %_2, align 1 
           48:  %1 = trunc i8 %0 to i1 
           49:  %2 = zext i1 %1 to i8 
           50:  call void @llvm.memset.p0.i64(ptr align 1 %x, i8 %2, i64 4, i1 false) 
           51: ; call slice_init::opaque 
           52:  call void @_ZN10slice_init6opaque17h4469190384ab35a6E(ptr align 1 %x) 
           53:  ret void 
           54: } 
           55:  
           56: ; Function Attrs: nonlazybind uwtable 
           57: define void @zeroed_integer_array() unnamed_addr #0 { 
           58: start: 
           59:  %x = alloca [16 x i8], align 4 
           60:  call void @llvm.memset.p0.i64(ptr align 4 %x, i8 0, i64 16, i1 false) 
           61: ; call slice_init::opaque 
           62:  call void @_ZN10slice_init6opaque17h587aa93d59c76bcaE(ptr align 4 %x) 
           63:  ret void 
           64: } 
           65:  
           66: ; Function Attrs: nonlazybind uwtable 
           67: define void @nonzero_integer_array() unnamed_addr #0 { 
           68: start: 
           69:  %x = alloca [16 x i8], align 4 
           70:  br label %repeat_loop_header 
           71:  
           72: repeat_loop_header: ; preds = %repeat_loop_body, %start 
           73:  %0 = phi i64 [ 0, %start ], [ %3, %repeat_loop_body ] 
           74:  %1 = icmp ult i64 %0, 4 
           75:  br i1 %1, label %repeat_loop_body, label %repeat_loop_next 
           76:  
           77: repeat_loop_body: ; preds = %repeat_loop_header 
           78:  %2 = getelementptr inbounds i32, ptr %x, i64 %0 
           79:  store i32 439041101, ptr %2, align 4 
           80:  %3 = add nuw i64 %0, 1 
           81:  br label %repeat_loop_header 
           82:  
           83: repeat_loop_next: ; preds = %repeat_loop_header 
           84: ; call slice_init::opaque 
           85:  call void @_ZN10slice_init6opaque17h587aa93d59c76bcaE(ptr align 4 %x) 
           86:  ret void 
           87: } 
           88:  
           89: ; Function Attrs: nonlazybind uwtable 
           90: define void @u16_init_one_bytes(ptr sret([200 x i8]) align 2 %_0) unnamed_addr #0 { 
           91: start: 
           92:  call void @llvm.memset.p0.i64(ptr align 2 %_0, i8 1, i64 200, i1 false) 
           93:  ret void 
           94: } 
           95:  
           96: ; Function Attrs: nonlazybind uwtable 
           97: define void @option_none_init(ptr sret([200 x i8]) align 1 %_0) unnamed_addr #0 { 
check:88'0                                  X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:88'0     ~~~~~~~
           99:  %_1 = alloca [2 x i8], align 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          100:  store i8 0, ptr %_1, align 1 
          100:  store i8 0, ptr %_1, align 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          101:  %0 = load i8, ptr %_1, align 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          102:  %1 = trunc i8 %0 to i1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~
          103:  %2 = getelementptr inbounds i8, ptr %_1, i64 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          104:  %3 = load i8, ptr %2, align 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          105:  br label %repeat_loop_header 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:88'0     ~
check:88'0     ~
          107: repeat_loop_header: ; preds = %repeat_loop_body, %start 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          108:  %4 = phi i64 [ 0, %start ], [ %9, %repeat_loop_body ] 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          109:  %5 = icmp ult i64 %4, 100 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
          110:  br i1 %5, label %repeat_loop_body, label %repeat_loop_next 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:88'0     ~
check:88'0     ~
          112: repeat_loop_body: ; preds = %repeat_loop_header 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          113:  %6 = getelementptr inbounds %"core::option::Option<u8>", ptr %_0, i64 %4 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          114:  %7 = zext i1 %1 to i8 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~
          115:  store i8 %7, ptr %6, align 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          116:  %8 = getelementptr inbounds i8, ptr %6, i64 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          117:  store i8 %3, ptr %8, align 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          118:  %9 = add nuw i64 %4, 1 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~
          119:  br label %repeat_loop_header 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:88'0     ~
check:88'0     ~
          121: repeat_loop_next: ; preds = %repeat_loop_header 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          122:  ret void 
check:88'0     ~~~~~~~~~~
          123: } 
check:88'0     ~~
check:88'0     ~
check:88'0     ~
          125: ; Function Attrs: nonlazybind uwtable 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          126: define void @option_maybe_uninit_init(ptr sret([200 x i8]) align 2 %_0) unnamed_addr #0 { 
check:88'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:88'1       ?                                                                                        possible intended match
          127: start: 
          128:  %0 = load i16, ptr @0, align 2 
          129:  br label %repeat_loop_header 
          130:  
          131: repeat_loop_header: ; preds = %repeat_loop_body, %start 
          132:  %1 = phi i64 [ 0, %start ], [ %4, %repeat_loop_body ] 
          133:  %2 = icmp ult i64 %1, 100 
          134:  br i1 %2, label %repeat_loop_body, label %repeat_loop_next 
          135:  
          136: repeat_loop_body: ; preds = %repeat_loop_header 
          137:  %3 = getelementptr inbounds i16, ptr %_0, i64 %1 
          138:  store i16 %0, ptr %3, align 2 
          139:  %4 = add nuw i64 %1, 1 
          140:  br label %repeat_loop_header 
          141:  
          142: repeat_loop_next: ; preds = %repeat_loop_header 
          143:  ret void 
          144: } 
          145:  
          146: ; Function Attrs: nonlazybind uwtable 
          147: define void @option_maybe_uninit_provenance(ptr sret([1200 x i8]) align 1 %_0) unnamed_addr #0 { 
          148: start: 
          149:  br label %repeat_loop_header 
          150:  
          151: repeat_loop_header: ; preds = %repeat_loop_body, %start 
          152:  %0 = phi i64 [ 0, %start ], [ %3, %repeat_loop_body ] 
          153:  %1 = icmp ult i64 %0, 100 
          154:  br i1 %1, label %repeat_loop_body, label %repeat_loop_next 
          155:  
          156: repeat_loop_body: ; preds = %repeat_loop_header 
          157:  %2 = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<Packed>", ptr %_0, i64 %0 
          158:  call void @llvm.memcpy.p0.p0.i64(ptr align 1 %2, ptr align 1 @1, i64 12, i1 false) 
          159:  %3 = add nuw i64 %0, 1 
          160:  br label %repeat_loop_header 
          161:  
          162: repeat_loop_next: ; preds = %repeat_loop_header 
          163:  ret void 
          164: } 
          166: ; slice_init::opaque 
          166: ; slice_init::opaque 
          167: ; Function Attrs: noinline nonlazybind uwtable 
          168: define void @_ZN10slice_init6opaque17h4469190384ab35a6E(ptr align 1 %_1) unnamed_addr #1 { 
          170:  ret void 
          171: } 
          172:  
          173: ; slice_init::opaque 
          173: ; slice_init::opaque 
          174: ; Function Attrs: noinline nonlazybind uwtable 
          175: define void @_ZN10slice_init6opaque17h587aa93d59c76bcaE(ptr align 4 %_1) unnamed_addr #1 { 
          177:  ret void 
          178: } 
          179:  
          180: ; slice_init::opaque 
          180: ; slice_init::opaque 
          181: ; Function Attrs: noinline nonlazybind uwtable 
          182: define void @_ZN10slice_init6opaque17h5a1c3b8d7de917daE(ptr align 1 %_1) unnamed_addr #1 { 
          184:  ret void 
          185: } 
          186:  
          187: ; slice_init::opaque 
          187: ; slice_init::opaque 
          188: ; Function Attrs: noinline nonlazybind uwtable 
          189: define void @_ZN10slice_init6opaque17h90ba94c9c1d53e83E(ptr align 4 %_1) unnamed_addr #1 { 
          191:  ret void 
          192: } 
          193:  
          194: ; slice_init::opaque 
          194: ; slice_init::opaque 
          195: ; Function Attrs: noinline nonlazybind uwtable 
          196: define void @_ZN10slice_init6opaque17h9c01bc0e6568e7f7E(ptr align 1 %_1) unnamed_addr #1 { 
          198:  ret void 
          199: } 
          200:  
          200:  
          201: ; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: write) 
          202: declare void @llvm.memset.p0.i64(ptr nocapture writeonly, i8, i64, i1 immarg) #2 
          203:  
          204: ; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite) 
          205: declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #3 
          206:  
          207: attributes #0 = { nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
          208: attributes #1 = { noinline nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
          209: attributes #2 = { nocallback nofree nounwind willreturn memory(argmem: write) } 
          210: attributes #3 = { nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
          211:  
          212: !llvm.module.flags = !{!0, !1} 
          213: !llvm.ident = !{!2} 
          214:  
          215: !0 = !{i32 8, !"PIC Level", i32 2} 
          216: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
          217: !2 = !{!"rustc version 1.86.0-nightly (663dd5e49 2025-01-21)"} 
------------------------------------------



```

</details>

--- a/compiler/rustc_codegen_gcc/src/common.rs
+++ b/compiler/rustc_codegen_gcc/src/common.rs
@@ -64,6 +64,11 @@ impl<'gcc, 'tcx> ConstCodegenMethods<'tcx> for CodegenCx<'gcc, 'tcx> {
         if type_is_pointer(typ) { self.context.new_null(typ) } else { self.const_int(typ, 0) }
     }
 
+    fn is_undef(&self, _val: RValue<'gcc>) -> bool {
+        // FIXME: actually check for undef
+        false
+    }
+
     fn const_undef(&self, typ: Type<'gcc>) -> RValue<'gcc> {
         let local = self.current_func.borrow().expect("func").new_local(None, typ, "undefined");
         if typ.is_struct().is_some() {

--- a/compiler/rustc_codegen_llvm/src/common.rs
+++ b/compiler/rustc_codegen_llvm/src/common.rs
@@ -126,6 +126,10 @@ impl<'ll, 'tcx> ConstCodegenMethods<'tcx> for CodegenCx<'ll, 'tcx> {
         unsafe { llvm::LLVMGetUndef(t) }
     }
 
+    fn is_undef(&self, v: &'ll Value) -> bool {
+        unsafe { llvm::LLVMIsUndef(v) == True }
+    }
+
     fn const_poison(&self, t: &'ll Type) -> &'ll Value {
         unsafe { llvm::LLVMGetPoison(t) }
     }

--- a/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
+++ b/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
@@ -918,6 +918,7 @@ unsafe extern "C" {
     pub fn LLVMMetadataTypeInContext(C: &Context) -> &Type;
 
     // Operations on all values
+    pub fn LLVMIsUndef(Val: &Value) -> Bool;
     pub fn LLVMTypeOf(Val: &Value) -> &Type;
     pub fn LLVMGetValueName2(Val: &Value, Length: *mut size_t) -> *const c_char;
     pub fn LLVMSetValueName2(Val: &Value, Name: *const c_char, NameLen: size_t);

--- a/compiler/rustc_codegen_ssa/src/mir/operand.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/operand.rs
@@ -204,14 +204,30 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         let alloc_align = alloc.inner().align;
         assert!(alloc_align >= layout.align.abi);
 
+        // Returns `None` when the value is partially undefined or any byte of it has provenance.
+        // Otherwise returns the value or (if the entire value is undef) returns an undef.
         let read_scalar = |start, size, s: abi::Scalar, ty| {
+            let range = alloc_range(start, size);
             match alloc.0.read_scalar(
                 bx,
-                alloc_range(start, size),
+                range,
                 /*read_provenance*/ matches!(s.primitive(), abi::Primitive::Pointer(_)),
             ) {
-                Ok(val) => bx.scalar_to_backend(val, s, ty),
-                Err(_) => bx.const_poison(ty),
+                Ok(val) => Some(bx.scalar_to_backend(val, s, ty)),
+                Err(_) => {
+                    // We may have failed due to partial provenance or unexpected provenance,
+                    // continue down the normal code path if so.
+                    if alloc.0.provenance().range_empty(range, &bx.tcx())
+                        // Since `read_scalar` failed, but there were no relocations involved, the
+                        // bytes must be partially or fully uninitialized. Thus we can now unwrap the
+                        // information about the range of uninit bytes and check if it's the full range.
+                        && alloc.0.init_mask().is_range_initialized(range).unwrap_err() == range
+                    {
+                        Some(bx.const_undef(ty))
+                    } else {
+                        None
+                    }
+                }
             }
         };
 
@@ -222,16 +238,14 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         // check that walks over the type of `mplace` to make sure it is truly correct to treat this
         // like a `Scalar` (or `ScalarPair`).
         match layout.backend_repr {
-            BackendRepr::Scalar(s @ abi::Scalar::Initialized { .. }) => {
+            BackendRepr::Scalar(s) => {
                 let size = s.size(bx);
                 assert_eq!(size, layout.size, "abi::Scalar size does not match layout size");
-                let val = read_scalar(offset, size, s, bx.immediate_backend_type(layout));
-                OperandRef { val: OperandValue::Immediate(val), layout }
+                if let Some(val) = read_scalar(offset, size, s, bx.immediate_backend_type(layout)) {
+                    return OperandRef { val: OperandValue::Immediate(val), layout };
+                }
             }
-            BackendRepr::ScalarPair(
-                a @ abi::Scalar::Initialized { .. },
-                b @ abi::Scalar::Initialized { .. },
-            ) => {
+            BackendRepr::ScalarPair(a, b) => {
                 let (a_size, b_size) = (a.size(bx), b.size(bx));
                 let b_offset = (offset + a_size).align_to(b.align(bx).abi);
                 assert!(b_offset.bytes() > 0);
@@ -247,20 +261,21 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
                     b,
                     bx.scalar_pair_element_backend_type(layout, 1, true),
                 );
-                OperandRef { val: OperandValue::Pair(a_val, b_val), layout }
+                if let (Some(a_val), Some(b_val)) = (a_val, b_val) {
+                    return OperandRef { val: OperandValue::Pair(a_val, b_val), layout };
+                }
             }
-            _ if layout.is_zst() => OperandRef::zero_sized(layout),
-            _ => {
-                // Neither a scalar nor scalar pair. Load from a place
-                // FIXME: should we cache `const_data_from_alloc` to avoid repeating this for the
-                // same `ConstAllocation`?
-                let init = bx.const_data_from_alloc(alloc);
-                let base_addr = bx.static_addr_of(init, alloc_align, None);
-
-                let llval = bx.const_ptr_byte_offset(base_addr, offset);
-                bx.load_operand(PlaceRef::new_sized(llval, layout))
-            }
+            _ if layout.is_zst() => return OperandRef::zero_sized(layout),
+            _ => {}
         }
+        // Neither a scalar nor scalar pair. Load from a place
+        // FIXME: should we cache `const_data_from_alloc` to avoid repeating this for the
+        // same `ConstAllocation`?
+        let init = bx.const_data_from_alloc(alloc);
+        let base_addr = bx.static_addr_of(init, alloc_align, None);
+
+        let llval = bx.const_ptr_byte_offset(base_addr, offset);
+        bx.load_operand(PlaceRef::new_sized(llval, layout))
     }
 
     /// Asserts that this operand refers to a scalar and returns

--- a/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
@@ -8,7 +8,7 @@ use rustc_middle::ty::{self, Instance, Ty, TyCtxt};
 use rustc_middle::{bug, mir, span_bug};
 use rustc_session::config::OptLevel;
 use rustc_span::{DUMMY_SP, Span};
-use tracing::{debug, instrument};
+use tracing::{debug, instrument, trace};
 
 use super::operand::{OperandRef, OperandValue};
 use super::place::PlaceRef;
@@ -93,6 +93,8 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     return;
                 }
 
+                // If `v` is an integer constant whose value is just a single byte repeated N times,
+                // emit a `memset` filling the entire `dest` with that byte.
                 let try_init_all_same = |bx: &mut Bx, v| {
                     let start = dest.val.llval;
                     let size = bx.const_usize(dest.layout.size.bytes());
@@ -117,13 +119,33 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     false
                 };
 
+                trace!(?cg_elem.val);
                 match cg_elem.val {
                     OperandValue::Immediate(v) => {
                         if try_init_all_same(bx, v) {
                             return;
                         }
                     }
-                    _ => (),
+                    OperandValue::Pair(a, b) => {
+                        let a_is_undef = bx.cx().is_undef(a);
+                        match (a_is_undef, bx.cx().is_undef(b)) {
+                            // Can happen for uninit unions
+                            (true, true) => {
+                                // FIXME: can we produce better output here?
+                            }
+                            (false, true) | (true, false) => {
+                                let val = if a_is_undef { b } else { a };
+                                if try_init_all_same(bx, val) {
+                                    return;
+                                }
+                            }
+                            (false, false) => {
+                                // FIXME: if both are the same value, use try_init_all_same
+                            }
+                        }
+                    }
+                    OperandValue::ZeroSized => unreachable!("checked above"),
+                    OperandValue::Ref(..) => {}
                 }
 
                 let count = self

--- a/compiler/rustc_codegen_ssa/src/traits/consts.rs
+++ b/compiler/rustc_codegen_ssa/src/traits/consts.rs
@@ -9,6 +9,7 @@ pub trait ConstCodegenMethods<'tcx>: BackendTypes {
     /// Generate an uninitialized value (matching uninitialized memory in MIR).
     /// Whether memory is initialized or not is tracked byte-for-byte.
     fn const_undef(&self, t: Self::Type) -> Self::Value;
+    fn is_undef(&self, v: Self::Value) -> bool;
     /// Generate a fake value. Poison always affects the entire value, even if just a single byte is
     /// poison. This can only be used in codepaths that are already UB, i.e., UB-free Rust code
     /// (including code that e.g. copies uninit memory with `MaybeUninit`) can never encounter a

--- a/compiler/rustc_middle/src/mir/interpret/allocation.rs
+++ b/compiler/rustc_middle/src/mir/interpret/allocation.rs
@@ -222,7 +222,7 @@ impl AllocError {
 }
 
 /// The information that makes up a memory access: offset and size.
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, PartialEq)]
 pub struct AllocRange {
     pub start: Size,
     pub size: Size,

--- a/tests/codegen/overaligned-constant.rs
+++ b/tests/codegen/overaligned-constant.rs
@@ -17,8 +17,6 @@ pub fn overaligned_constant() {
     // CHECK-LABEL: @overaligned_constant
     // CHECK: [[full:%_.*]] = alloca [32 x i8], align 8
     // CHECK: call void @llvm.memcpy.p0.p0.i64(ptr align 8 [[full]], ptr align 8 @0, i64 32, i1 false)
-    // CHECK: %b.0 = load i32, ptr @0, align 4
-    // CHECK: %b.1 = load i32, ptr getelementptr inbounds ({{.*}}), align 4
     let mut s = S(1);
 
     s.0 = 3;

--- a/tests/codegen/slice-init.rs
+++ b/tests/codegen/slice-init.rs
@@ -2,6 +2,8 @@
 
 #![crate_type = "lib"]
 
+use std::mem::MaybeUninit;
+
 // CHECK-LABEL: @zero_sized_elem
 #[no_mangle]
 pub fn zero_sized_elem() {
@@ -76,16 +78,14 @@ pub fn u16_init_one_bytes() -> [u16; N] {
     [const { u16::from_be_bytes([1, 1]) }; N]
 }
 
-// FIXME: undef bytes can just be initialized with the same value as the
-// defined bytes, if the defines bytes are all the same.
 // CHECK-LABEL: @option_none_init
 #[no_mangle]
 pub fn option_none_init() -> [Option<u8>; N] {
     // CHECK-NOT: select
-    // CHECK: br label %repeat_loop_header{{.*}}
+    // CHECK-NOT: br
     // CHECK-NOT: switch
-    // CHECK: icmp
-    // CHECK-NOT: call void @llvm.memset.p0
+    // CHECK-NOT: icmp
+    // CHECK: call void @llvm.memset.p0
     [const { None }; N]
 }
 

--- a/tests/codegen/slice-init.rs
+++ b/tests/codegen/slice-init.rs
@@ -86,7 +86,7 @@ pub fn option_none_init() -> [Option<u8>; N] {
     // CHECK-NOT: switch
     // CHECK: icmp
     // CHECK-NOT: call void @llvm.memset.p0
-    [None; N]
+    [const { None }; N]
 }
 
 // Use an opaque function to prevent rustc from removing useless drops.

